src/main/java/dev/latvian/mods/kubejs/neoforge/NativeEvents.java

```diff
@@ -0,0 +1,147 @@
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class NativeEvents implements CustomJavaToJsWrapper {
+
+	public static final String NAME = "NativeEvents";
+	@Nullable
+	private static NativeEvents lastKnownBus;
+	private final Map<String, NativeEventHandler> handlers = new HashMap<>();
+	private final List<NativeEventConsumer> directConsumers = new ArrayList<>();
+
+	public static NativeEvents create() {
+		if (lastKnownBus != null) {
+			lastKnownBus.clear();
+		}
+
+		NativeEvents bus = new NativeEvents();
+		lastKnownBus = bus;
+		return bus;
+	}
+
+	private void clear() {
+		ScriptType.STARTUP.console.info("Clearing native events");
+		for (var handler : handlers.values()) {
+			handler.unregister();
+		}
+		handlers.clear();
+
+		for (var consumer : directConsumers) {
+			NeoForge.EVENT_BUS.unregister(consumer);
+		}
+		directConsumers.clear();
+	}
+
+	private NativeEvents() {
+	}
+
+	@Override
+	public Scriptable convertJavaToJs(Context context, Scriptable scriptable, Class<?> aClass) {
+		return new Wrapper(scriptable, this, aClass, context);
+	}
+
+	@Nullable
+	public NativeEventHandler getHandler(String name) {
+		var handler = handlers.get(name);
+		if (handler != null) {
+			return handler;
+		}
+
+		var eventClass = NeoForgeEventsLookup.INSTANCE.get(name);
+		if (eventClass == null) {
+			return null;
+		}
+
+		handler = new NativeEventHandler(name, eventClass);
+		handlers.put(name, handler);
+		return handler;
+	}
+
+	public void printAllEvents() {
+		NeoForgeEventsLookup.INSTANCE.getEvents()
+			.entrySet()
+			.stream()
+			.sorted(Comparator.comparing(Map.Entry::getKey))
+			.forEach(entry -> KubeJS.LOGGER.info(entry.getKey() + " - ['" + entry.getValue().getName() + "']"));
+	}
+
+	public Object onEvent(Object eventClass, NativeEventConsumer consumer) {
+		return onEvent(EventPriority.NORMAL, eventClass, consumer);
+	}
+
+	public Object onEvent(EventPriority priority, Object eventClass, NativeEventConsumer consumer) {
+		if (!(eventClass instanceof CharSequence || eventClass instanceof Class)) {
+			throw new RuntimeException("Invalid syntax! " + NAME + ".onEvent(eventType, function) requires event class and handler");
+		}
+
+		try {
+			var type = eventClass instanceof Class<?> c ? c : Class.forName(eventClass.toString());
+			var secured = secure(consumer);
+			//noinspection unchecked
+			NeoForge.EVENT_BUS.addListener(priority, false, (Class<Event>) type, secured);
+			directConsumers.add(secured);
+		} catch (Exception ex) {
+			throw new RuntimeException(ex);
+		}
+
+		return null;
+	}
+
+	private NativeEventConsumer secure(NativeEventConsumer consumer) {
+		return event -> {
+			try {
+				consumer.accept(event);
+			} catch (Exception ex) {
+				ScriptType.STARTUP.console.error(ex.getLocalizedMessage(), ex);
+				KubeJS.LOGGER.error(ex.getLocalizedMessage(), ex);
```

Comment:
This should have more information about the event handler, like the errors for the callbacks in custom blocksz

Reply:
Sounds good. Will do